generative-ai: Add anti pattern example (#1679)
python/devguide: getting-started/generative-ai.rst

Generative AI
=============

> Generative AI has evolved rapidly and can be impressive. As with using any tool, the resulting contribution is

Reviewer: ```suggestion
Generative AI tools have evolved rapidly, and their suggested results can be helpful. As with using any tool, the resulting contribution is
```

Maintainers may close issues and PRs that are not useful or productive, including
those that are fully generated by AI. If a contributor repeatedly opens unproductive
issues or PRs, they may be blocked.

Anti-patterns
=============

> - While AI-assisted tools such as autocompletion can enhance productivity, they sometimes rewrite entire code blocks instead of making small, focused edits.

Reviewer 2: I'd actually tighten this phrasing up a bit.  I find this list of anti-patterns to not be AI specific.  We already reject PRs coming from people who make larger than necessary changes that are difficult to review.  So I'd suggest wording it as a "positive desired thing, not negative example" phrasing.

- Make small focused edits, prefer not rewriting entire code blocks.
   This makes it easier to review changes and fully understand both the original intent of the code and the rationale behind the new modifications.
   ...

  This can make it more difficult to review changes and to fully understand both the original intent of the code and the rationale behind the new modifications.
  Maintaining consistency with the original code helps preserve clarity, traceability, and meaningful reviews and also helps us avoid unnecessary code churn.
- Sometimes AI assisted tools make failing unit tests pass by altering or bypassing the tests rather than addressing the underlying problem in the code.

>   Such changes do not represent a real fix and should be avoided.

Reviewer 2: I think this sentence was only intended to be tied to the false testing / false success bullet it is part of.  Rather than wording it as "should be avoided" the right thing for users to do is to request that authors have reviewed the work in detail for problems like this before proposing it as a PR.

I'd focus on a proactive positive outcome based wording rather than a negative like "avoid" or "discourage".

The reason being, the statements in here don't reflect all AI tooling. This devguide doc is about the use of AI, so we should assume those who read it are interested in or already using AI and are here to learn how to do it in an appropriate and productive manner.

Statements like "we discourage use of AI when ..." are likely to be outdated very quickly and discourage things that won't be true for long.  Thus my desire to focus this on what we do want and why it is important for project health and reviewer respect.

Maintainers may close issues and PRs that are not useful or productive, including
those that are fully generated by AI. If a contributor repeatedly opens unproductive
issues or PRs, they may be blocked.

> Considerations or Keys for Success

Reviewer 3: I think a simpler heading works better:
```suggestion
Considerations for success
```

> ==================================

Reviewer 4: Use sentence case for headers:

https://devguide.python.org/documentation/style-guide/#capitalization
```suggestion
Considerations or keys for success
==================================

```

- While AI assisted tools such as autocompletion can enhance productivity, they sometimes rewrite entire code blocks instead of making small, focused edits.
  This can make it more difficult to review changes and to fully understand both the original intent of the code and the rationale behind the new modifications.
  Maintaining consistency with the original code helps preserve clarity, traceability, and meaningful reviews and also helps us avoid unnecessary code churn.
- Sometimes AI assisted tools make failing unit tests pass by altering or bypassing the tests rather than addressing the underlying problem in the code.
  Such changes do not represent a real fix. Authors must review the work done by AI tooling in detail to ensure it actually makes sense before proposing it as a PR.
- Please keep following principles for the quality of your contributions in mind whether you use generative AI or not:
  - Consider whether the change is necessary.
  - Make minimal, focused changes.
  - Follow existing coding style and patterns.

>   - Write tests that exercise the change.

Reviewer 4: Added to the suggestion.

Maintainers may close issues and PRs that are not useful or productive, including
those that are fully generated by AI. If a contributor repeatedly opens unproductive
issues or PRs, they may be blocked.

> Anti-patterns

Reviewer 2: i think you covered the incorrectly changing tests instead of fixing one. Good call. I like having that in the list.

I don't personally see that so often anymore with Opus 4.1 or Sonnet 4.5, but older or lower end models tend to reward hack their way into that state more often - it's a thing we'll likely encounter coming from new contributors for a while as they can't be assumed to be using the latest and greatest.
